### ISSUE TITLE
String matchers support for non-standard string classes

### ISSUE THREAD

**Issue by qtiki** [google/googletest#1050]
I've recently updated from gmock-1.5.0 to googletest-1.8.0. We use our own String class which used to work with testing::HasSubstr and testing::StrEq (etc..) matchers. After updating I get type-casting error trying to convert from our own String to const testing::internal::string& with all of these matchers.

I narrowed this problem down to a few places in gmock-matchers.h. For example in StrEqualityMatcher:

```template <typename MatcheeStringType>
bool MatchAndExplain(const MatcheeStringType& s,
                     MatchResultListener* /* listener */) const {
  const StringType& s2(s);
  const bool eq = case_sensitive_ ? s2 == string_ :
    CaseInsensitiveStringEquals(s2, string_);
  return expect_eq_ == eq;
```

Changing the `const StringType& s2(s)` reference to a copy-constructor `const StringType s2(s)` would allow using any custom string classes that can be implicitly converted to types accepted by testing::internal::string constructor (such as const char*). Naturally it would be a bit slower with some extra data being moved but it might be worth it. If the performance hit is too scary it might be possible to use some template magic to do this copying only where the reference is not possible.

**Comment by BillyDonahue** [comment]
I don't understand what's going wrong for you. `const StringType& s2(s);` should already be capable of creating a temporary s2 and binding it to `s2` with temporary lifetime extension.

```
struct X{} x;
struct Y{Y(X);};
void f() {
  const Y& y(x);
}
```

**Comment by tribetiki** [comment]
I tried your example and it does indeed compile. However the case with our String class is slightly different as it can't directly be passed to the testing::internal::string constructor but needs to use it's custom const char* casting operator. Here's your example modified to illustrate the issue:

```
struct X{operator int();} x;
struct Y{Y(int);};
void f() { const Y& y(x); }
```

Trying to compile that (at least with Visual Studio 2012) produces this:

```
error C2440: 'initializing' : cannot convert from 'X' to 'const Y &'
Reason: cannot convert from 'X' to 'const Y'
No user-defined-conversion operator available that can perform this conversion,
 or the operator cannot be called
```

Modifying the example to `void f() { const Y y(x); }` will make it compile without errors.

**Comment by BillyDonahue** [comment]
Sorry why are we talking about `int` in your revised code?

```
struct X{operator int();} x;
struct Y{Y(int);};
void f() { const Y& y(x); }
```

You should be able to define a conversion from your string to std::string.
```
#include <string>
struct YourString { operator std::string() const; };
void f(const YourString& x) { const std::string& y(x); }
```

**Comment by tribetiki** [comment]
The int was just the shortest type to demonstrate the casting issue in minimum amount of space.

I actually tried the std::string conversion first but it generated a lot of compile errors with the const char* casting operator we also have, mainly ambiguous conversions all around. Also the std::strings we use in our code have a custom memory allocator so as far as types go they're not directly compatible with std::string. The codebase is rather big and is no longer actively worked on so I did not want to spend too much time fixing it - especially considering that it used to work with gmock-1.5.0. I only updated googletest because of some OS X specific compile issues after our OS X target platform was updated. Easiest fix for us was to modify googletest code directly - I just figured I'd post this here in case you'd want to support non-standard string types with only const char* operators (and I won't hold it against you if you don't...).
